app/models/katello/repository.rb

```diff
@@ -483,8 +468,37 @@ def import_system_applicability
       end
     end
 
+    def units_for_removal(ids)
+      table_name = removable_unit_association.table_name
+      is_integer = Integer(ids.first) rescue false #assume all ids are either integers or not
+
+      if is_integer
+        self.removable_unit_association.where("#{table_name}.id in (?)", ids)
+      else
+        self.removable_unit_association.where("#{table_name}.uuid in (?)", ids)
+      end
+    end
+
+    def remove_db_units(units)
+      if yum?
+        self.rpms -= units
+      elsif docker?
+        remove_docker_db_units(units)
+      end
+    end
+
     protected
 
+    def removable_unit_association
+      if yum?
+        self.rpms
+      elsif docker?
+        self.docker_images
+      else
+        fail "Data type not supported"
```

Comment:
It really should never bubble up to the user, and I Image once you add puppet you'll remove it